### DISCUSSION TITLE
Unable to merge object hierarchy

### DISCUSSION THREAD

**Discussion by rcdailey** [LuckyPennySoftware/AutoMapper#4228] [NA]
I have two objects:

- A business object that was deserialized from JSON data and used by my application (`Foo1`)
- A DTO object created from `Foo1` sent to a remote REST API and that is mostly similar but not entirely (`Foo2`).

Using AutoMapper, I want to "merge" `Foo1` into `Foo2`. One thing I think that makes this tricky is that `Foo2` expects a *list* of `NestedObject2`, whereas `Foo1` only has a single `NestedObject1` instead of a list. Due to the way the API works, `NestedObject2` will *never* have more than one item in it. I put more than one in my unit test only to show that `"one"` gets replaced (updated) and `"two"` doesn't get clobbered (to verify merge behavior).

Below is a unit test with example objects that represent what I'm trying to do. The problem is that `AfterMap` gives me a destination dictionary (`ExtraObjects`) that is empty. I'm not sure why. Honestly, the whole way I've set up the mapping profile just "feels" wrong to me. I don't feel confident that I'm doing this right.

How should I get this working?

The objects:

```cs
public record NestedObject1
{
    public Dictionary<string, object> ExtraObjects { get; init; } = new();
}

public record NestedObject2
{
    public Dictionary<string, object> ExtraObjects { get; init; } = new();
}

public record Foo1
{
    public NestedObject1 Objects { get; set; } = new();
}

public record Foo2
{
    public IReadOnlyCollection<NestedObject2> Objects { get; set; } = Array.Empty<NestedObject2>();
}
```

The mapping profile:

```cs
public class ExampleMappingProfile : Profile
{
    public ExampleMappingProfile()
    {
        CreateMap<NestedObject1, NestedObject2>();

        CreateMap<Foo1, Foo2>()
            .ForMember(d => d.Objects, o => o.Ignore())
            .AfterMap((s, d, ctx) =>
            {
                if (d.Objects.IsEmpty())
                {
                    d.Objects = new [] {ctx.Mapper.Map<NestedObject2>(s.Objects)};
                }
                else
                {
                    ctx.Mapper.Map(s.Objects, d.Objects.First());
                }
            });
    }
}
```

And the unit test:

```cs
[Test]
public void Example()
{
    var source = new Foo1
    {
        Objects = new NestedObject1
        {
            ExtraObjects =
            {
                {"one", 1}
            }
        }
    };

    var dest = new Foo2
    {
        Objects = new[]
        {
            new NestedObject2
            {
                ExtraObjects =
                {
                    {"one", 100},
                    {"two", 2}
                }
            }
        }
    };

    var mapper = Resolve<IMapper>();

    mapper.Map(source, dest);

    dest.Should().BeEquivalentTo(new Foo2
    {
        Objects = new[]
        {
            new NestedObject2
            {
                ExtraObjects =
                {
                    {"one", 1},
                    {"two", 2}
                }
            }
        }
    });
}
```

The result of the test is:

> Expected property dest.Objects[0].ExtraObjects to be a dictionary with 2 item(s), but it misses key(s) {"two"}

**Comment by lbargaoanu** [comment]
I think this is better suited for StackOverflow.

**Reply by rcdailey** [reply]
Funny, I posted this there first and I got downvoted like crazy. So I'm keeping it here.

**Reply by jbogard** [reply]
That might be a sign 🤣

AutoMapper is not a library designed for merging object hierarchies. Maybe it will work. Probably it won't. I would pick a different solution.

**Reply by rcdailey** [reply]
Well the only other solution I can think of is to use Newtonsoft.Json to merge (`Populate`) them and then add a special case to handle the single item -> list. But I assumed it would be cleaner to do it in AutoMapper.

If AutoMapper is not designed for merging objects, then what is the overload of `Map()` for that takes a source & dest? Am I misunderstanding the purpose despite having read the documentation?

**Reply by jbogard** [reply]
The overload for Map that takes a source/dest is for mapping into an existing instance. It won't replace references but it will wipe everything else out. It's for cases where you have ORM-specific proxy objects/tracked objects and replacing references messes up the ORM.
